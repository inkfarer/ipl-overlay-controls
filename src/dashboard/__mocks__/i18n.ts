import i18next from 'i18next';
import { config } from '@vue/test-utils';

jest.mock('i18next-vue', () => ({
    __esModule: true,
    useTranslation: () => ({ t: i18next.t })
}));

beforeAll(async () => {
    await i18next.init({
        lng: 'cimode',
        appendNamespaceToCIMode: true
    });
    config.global.mocks = {
        $t: i18next.t
    };
});
