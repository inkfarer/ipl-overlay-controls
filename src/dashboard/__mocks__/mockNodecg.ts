import Mock = jest.Mock;

export type ReplicantChangeHandler = (newValue?: unknown, oldValue?: unknown) => void;
export type MockReplicant = { value: unknown };
type OnFunction = (event: string, handler: ReplicantChangeHandler) => void;

export class MockNodecg {
    replicants: {[key: string]: MockReplicant};
    listeners: {[key: string]: ReplicantChangeHandler};
    sendMessage: Mock

    constructor() {
        this.replicants = {};
        this.listeners = {};
        this.sendMessage = jest.fn();
    }

    init(): void {
        const self = this;
        window.nodecg = {
            // @ts-ignore: Just a mock
            Replicant(name: string) {
                const replicantValue: { value: unknown, on: OnFunction } = {
                    value: undefined,
                    on: (event: string, handler: ReplicantChangeHandler) => {
                        self.listeners[name] = handler;
                    }
                };
                self.replicants[name] = replicantValue;
                return replicantValue;
            },
            sendMessage: self.sendMessage,
            // @ts-ignore: Just a mock
            log: {
                info: jest.fn()
            }
        };

        // @ts-ignore: It's a mock.
        window.NodeCG = {
            waitForReplicants: jest.fn().mockResolvedValue(true)
        };
    }
}
