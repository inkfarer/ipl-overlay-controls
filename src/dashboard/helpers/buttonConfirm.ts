/*
buttonConfirm

After importing this script to a dashboard panel, add the attribute 'data-requires-confirmation' to any button.
The button will now ask for confirmation when it is first clicked.
When it is clicked a second time, a 'confirm' event is sent by it, which can be listened for by scripts.

Example:

HTML:
    <button id="dangerous-action-button" data-requires-confirmation>DO A THING</button>
JS:
    document.getElementById('dangerous-action-button').addEventListener('confirm', () => {
        // do something
    });
*/

window.addEventListener('DOMContentLoaded', () => {
    function addListener(btn: HTMLButtonElement) {
        if (btn.dataset.hasConfirmListener !== 'true') {
            btn.addEventListener('click', e => {
                const target = e.target as HTMLButtonElement;

                function resetButton() {
                    target.dataset.isClicked = 'false';
                    target.innerText = target.dataset.originalText;
                    target.style.backgroundColor = '';
                }

                if (target.dataset.isClicked === 'true') {
                    target.dispatchEvent(new Event('confirm'));
                    resetButton();
                } else {
                    target.dataset.isClicked = 'true';
                    target.dataset.originalText = target.innerText;
                    target.style.backgroundColor = 'var(--red)';
                    target.innerText = 'are you sure?';

                    setTimeout(() => {
                        resetButton();
                    }, 5000);
                }
            });

            btn.dataset.hasConfirmListener = 'true';
        }
    }

    const buttons = document.querySelectorAll('button[data-requires-confirmation]');
    buttons.forEach(addListener);

    const observer = new MutationObserver(mutations => {
        mutations.forEach(mutation => {
            mutation.addedNodes.forEach(node => {
                if (node.nodeType === Node.ELEMENT_NODE
                    && node.nodeName === 'BUTTON'
                    && (node as HTMLElement).dataset.requiresConfirmation) {

                    const btn = node as HTMLButtonElement;
                    addListener(btn);
                }
            });
        });
    });

    observer.observe(document.body, {
        subtree: true,
        childList: true
    });
});
