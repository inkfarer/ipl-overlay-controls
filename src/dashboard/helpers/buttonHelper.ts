import { colors } from '../styles/colors';

export function setToggleButtonDisabled(
    toggleTrue: HTMLButtonElement,
    toggleFalse: HTMLButtonElement,
    state: boolean
): void {
    toggleTrue.disabled = state;
    toggleFalse.disabled = !state;
}

export function addChangeReminder(
    elements: HTMLElement[] | NodeListOf<HTMLElement>,
    updateButton: HTMLButtonElement
): void {
    elements.forEach(element => {
        if (!element.tagName) return;

        let event;
        if (element.tagName.toLowerCase() === 'input') event = 'input';
        else if (element.tagName.toLowerCase() === 'select') event = 'change';
        else return;

        element.addEventListener(event, e => {
            const target = e.target as HTMLElement;
            target.dataset.edited = 'true';
            updateButton.style.backgroundColor = colors.red;
        });
    });

    updateButton.addEventListener('click', () => {
        elements.forEach(elem => {
            elem.dataset.edited = 'false';
        });
        updateButton.style.backgroundColor = colors.blue;
    });
}
