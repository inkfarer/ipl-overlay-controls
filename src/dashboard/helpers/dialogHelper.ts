import { NodecgDialog } from '../types/dialog';

export function closeDialog(dialogName: string): void {
    (nodecg.getDialog(dialogName) as NodecgDialog).close();
}
