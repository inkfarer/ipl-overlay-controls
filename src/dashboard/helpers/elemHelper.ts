export function appendChildren(element: HTMLElement, ...children: HTMLElement[]): void {
    children.forEach(elem => {
        element.appendChild(elem);
    });
}

export function elementById<T extends HTMLElement>(id: string): T | undefined {
    return document.getElementById(id) as T;
}

export function dispatch(element: EventTarget, event: string): void {
    element.dispatchEvent(new Event(event));
}

export function hideElement(element: HTMLElement): void {
    element.style.display = 'none';
}

export function showElement(element: HTMLElement): void {
    element.style.removeProperty('display');
}

export function prependToBody(elem: HTMLElement): void {
    document.body.prepend(elem);
}

export function appendElementAfter(newElement: HTMLElement, element: HTMLElement): void {
    element.parentNode.insertBefore(newElement, element.nextSibling);
}
