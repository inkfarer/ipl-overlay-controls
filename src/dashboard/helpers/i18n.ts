import { RuntimeConfig } from 'schemas';
import i18next, { type BackendModule } from 'i18next';

const I18nBackend: BackendModule = {
    type: 'backend',

    // eslint-disable-next-line @typescript-eslint/no-empty-function
    init() {

    },

    async read(language, namespace) {
        return loadTranslation(language, namespace);
    },
};

export async function initI18n(namespaces: string | string[]): Promise<void> {
    const ns = Array.isArray(namespaces) ? [...namespaces, 'common']: [namespaces, 'common'];
    await i18next
        .use(I18nBackend)
        .init({
            lng: 'en',
            fallbackLng: 'en',
            ns,
            interpolation: {
                escapeValue: false
            }
        });

    const runtimeConfig = nodecg.Replicant<RuntimeConfig>('runtimeConfig');
    runtimeConfig.on('change', async (newValue, oldValue) => {
        if (!oldValue || newValue.interfaceLocale !== oldValue.interfaceLocale) {
            await i18next.changeLanguage(newValue.interfaceLocale.toLowerCase()).catch(e => {
                console.error('Failed to change interface language', e);
            });
            if (i18next.exists('title')) {
                try {
                    window.frameElement.parentElement.setAttribute('display-title', i18next.t('title'));
                } catch (ignore) {

                }
            }
        }
    });
}

async function loadTranslation(locale: string, name: string) {
    try {
        return await import(`../../helpers/i18n/${locale.toLowerCase()}/${name}.json`);
    } catch (ignore) {
        return null;
    }
}
