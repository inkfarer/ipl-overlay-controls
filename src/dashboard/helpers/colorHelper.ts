import { colors, getColorOptionName } from '../../helpers/splatoonData';

export function fillColorSelector(select: HTMLSelectElement): void {
    for (let i = 0; i < colors.length; i++) {
        const element = colors[i];

        const optGroup = document.createElement('optgroup');
        optGroup.label = element.meta.name;

        for (let j = 0; j < element.colors.length; j++) {
            const color = element.colors[j];

            const option = document.createElement('option');
            option.value = getColorOptionName(color.index, element.meta.name);
            option.text = color.title;
            option.dataset.index = String(color.index);
            option.dataset.firstColor = color.clrA;
            option.dataset.secondColor = color.clrB;
            option.dataset.categoryName = element.meta.name;
            option.disabled = color.isCustom;

            optGroup.appendChild(option);
        }

        select.appendChild(optGroup);
    }
}

export function getContrastingTextColor(backgroundColor: string): string {
    const { r, g, b } = hexToRbg(backgroundColor);
    const yiq = ((r * 299) + (g * 587) + (b * 114)) / 1000;
    return (yiq >= 128) ? '#333' : 'white';
}

export function hexToRbg(color: string): { r: number, g: number, b: number } {
    color = color.replace('#', '');
    return {
        r: parseInt(color.substr(0, 2), 16),
        g: parseInt(color.substr(2, 2), 16),
        b: parseInt(color.substr(4, 2), 16)
    };
}
