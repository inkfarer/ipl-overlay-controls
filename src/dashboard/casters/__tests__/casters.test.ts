import { MockNodecg } from '../../__mocks__/mockNodecg';
import { appendChildren, dispatch, elementById } from '../../helpers/elemHelper';
import { Module } from '../../../helpers/__mocks__/module';
import { Casters } from 'schemas';

describe('casters', () => {
    const mockGenerateId = jest.fn();
    let nodecg: MockNodecg;
    let casters: Module;

    jest.mock('../../../helpers/generateId', () => ({
        __esModule: true,
        generateId: mockGenerateId
    }));

    beforeEach(() => {
        jest.resetModules();
        jest.resetAllMocks();

        nodecg = new MockNodecg();
        nodecg.init();

        document.body.innerHTML = `
            <div id="casters"></div>
            <button id="add-caster-btn"></button>`;

        casters = require('../casters');
    });

    describe('casters: change', () => {
        it('adds casters', () => {
            nodecg.listeners.casters({
                casterone: {
                    name: 'Caster the First',
                    twitter: '@caster1',
                    pronouns: 'he/him'
                },
                castertwo: {
                    name: 'Caster Two',
                    twitter: '@caster2',
                    pronouns: 'she/her'
                }
            });

            expect(elementById('casters').innerHTML).toMatchSnapshot();
        });

        it('handles casters being added and updated', () => {
            const casterElem = elementById('casters');

            const initialCasters = {
                casterone: {
                    name: 'Caster the First',
                    twitter: '@caster1',
                    pronouns: 'he/him'
                },
                castertwo: {
                    name: 'Caster Two',
                    twitter: '@caster2',
                    pronouns: 'she/her'
                }
            };
            nodecg.listeners.casters(initialCasters);

            expect(casterElem.innerHTML).toMatchSnapshot();

            const newCasters = {
                casterone: {
                    name: 'Caster the First (Updated)',
                    twitter: '@caster1new',
                    pronouns: 'they/them'
                },
                castertwo: {
                    name: 'Caster Two',
                    twitter: '@caster2',
                    pronouns: 'she/her'
                },
                casterthree: {
                    name: 'Third Caster',
                    twitter: '@casterthree',
                    pronouns: 'they/them'
                }
            };
            nodecg.listeners.casters(newCasters, initialCasters);

            expect(casterElem.innerHTML).toMatchSnapshot();
        });

        it('handles casters being removed', () => {
            const casterElem = elementById('casters');

            const initialCasters = {
                casterone: {
                    name: 'Caster the First',
                    twitter: '@caster1',
                    pronouns: 'he/him'
                },
                castertwo: {
                    name: 'Caster Two',
                    twitter: '@caster2',
                    pronouns: 'she/her'
                }
            };
            nodecg.listeners.casters(initialCasters);

            expect(casterElem.innerHTML).toMatchSnapshot();

            const newCasters = {
                casterone: {
                    name: 'Caster the First',
                    twitter: '@caster1',
                    pronouns: 'he/him'
                }
            };
            nodecg.listeners.casters(newCasters, initialCasters);

            expect(casterElem.innerHTML).toMatchSnapshot();
        });

        it('disables uncommitted casters if the limit for casters is reached', () => {
            /*
            * - At first, there are two casters
            * - The user creates a third caster but does not save it
            * - The casters replicant is updated with a third caster (for example, from another dashboard user)
            * - The user should not be able to save the caster they just created
            */

            const initialCasters = {
                casterone: {
                    name: 'Caster the First',
                    twitter: '@caster1',
                    pronouns: 'he/him'
                },
                castertwo: {
                    name: 'Caster Two',
                    twitter: '@caster2',
                    pronouns: 'she/her'
                }
            };
            nodecg.listeners.casters(initialCasters);

            casters.updateOrCreateCreateCasterElem('newcaster', undefined, true);

            const newCasters = {
                casterone: {
                    name: 'Caster the First (Updated)',
                    twitter: '@caster1new',
                    pronouns: 'they/them'
                },
                castertwo: {
                    name: 'Caster Two',
                    twitter: '@caster2',
                    pronouns: 'she/her'
                },
                casterthree: {
                    name: 'Third Caster',
                    twitter: '@casterthree',
                    pronouns: 'they/them'
                }
            };
            nodecg.listeners.casters(newCasters, initialCasters);

            expect(elementById<HTMLButtonElement>('update-caster_newcaster').disabled).toEqual(true);
        });

        it('disables the create caster button if we have hit the caster limit', () => {
            nodecg.listeners.casters({
                casterone: {
                    name: 'Caster the First (Updated)',
                    twitter: '@caster1new',
                    pronouns: 'they/them'
                },
                castertwo: {
                    name: 'Caster Two',
                    twitter: '@caster2',
                    pronouns: 'she/her'
                },
                casterthree: {
                    name: 'Third Caster',
                    twitter: '@casterthree',
                    pronouns: 'they/them'
                }
            });

            expect(elementById<HTMLButtonElement>('add-caster-btn').disabled).toEqual(true);
        });
    });

    describe('add-caster-btn: click', () => {
        it('creates a caster', () => {
            mockGenerateId.mockReturnValue('idid123');
            const casterElem = elementById('casters');
            casterElem.innerHTML = '';

            dispatch(elementById('add-caster-btn'), 'click');

            expect(elementById('caster-container_idid123')).not.toBeNull();
        });

        it('disables the button if the caster limit has been reached', () => {
            const casterElem = elementById('casters');
            casterElem.innerHTML = '<div class="caster-container"></div><div class="caster-container"></div>';
            const button = elementById<HTMLButtonElement>('add-caster-btn');

            dispatch(button, 'click');

            expect(button.disabled).toEqual(true);
        });
    });

    describe('updateOrCreateCasterElem', () => {
        it('creates element with data if it does not exist', () => {
            const casterElem = elementById('casters');
            casterElem.innerHTML = '';

            casters.updateOrCreateCreateCasterElem(
                'newcaster', { name: 'new caster', twitter: '@casternew', pronouns: 'she/her' }, true);

            expect(elementById('caster-container_newcaster')).not.toBeUndefined();
            expect(elementById<HTMLInputElement>('caster-name-input_newcaster').value).toEqual('new caster');
            expect(elementById<HTMLInputElement>('caster-twitter-input_newcaster').value).toEqual('@casternew');
            expect(elementById<HTMLInputElement>('caster-pronoun-input_newcaster').value).toEqual('she/her');
            expect(elementById('update-caster_newcaster').classList).toContain('uncommitted');
        });

        it('creates element with data if it does not exist and is committed', () => {
            const casterElem = elementById('casters');
            casterElem.innerHTML = '';

            casters.updateOrCreateCreateCasterElem(
                'newcaster', { name: 'new caster', twitter: '@casternew', pronouns: 'she/her' }, false);

            expect(elementById('caster-container_newcaster')).not.toBeUndefined();
            expect(elementById<HTMLInputElement>('caster-name-input_newcaster').value).toEqual('new caster');
            expect(elementById<HTMLInputElement>('caster-twitter-input_newcaster').value).toEqual('@casternew');
            expect(elementById<HTMLInputElement>('caster-pronoun-input_newcaster').value).toEqual('she/her');
            expect(elementById('update-caster_newcaster').classList).not.toContain('uncommitted');
        });

        it('creates event listener for new update caster button', () => {
            const casterElem = elementById('casters');
            casterElem.innerHTML = '';
            nodecg.replicants.casters.value = {};

            casters.updateOrCreateCreateCasterElem(
                'newcaster', { name: 'new caster', twitter: '@casternew', pronouns: 'she/her' }, true);

            const updateButton = elementById('update-caster_newcaster');
            expect(updateButton.classList).toContain('uncommitted');

            dispatch(updateButton, 'click');

            expect((nodecg.replicants.casters.value as Casters)['newcaster']).toEqual({
                name: 'new caster',
                twitter: '@casternew',
                pronouns: 'she/her'
            });
            expect(updateButton.classList).not.toContain('uncommitted');
        });

        describe('event listener for new delete caster button', () => {
            it('updates replicant if caster exists', () => {
                const casterElem = elementById('casters');
                casterElem.innerHTML = '';
                nodecg.replicants.casters.value = {
                    newcaster: { name: 'new caster' }
                };

                casters.updateOrCreateCreateCasterElem(
                    'newcaster', { name: 'new caster', twitter: '@casternew', pronouns: 'she/her' }, true);

                const removeButton = elementById('remove-caster_newcaster');
                dispatch(removeButton, 'click');

                expect((nodecg.replicants.casters.value as Casters)['newcaster']).toBeUndefined();
            });

            it('removes element if caster is not committed', () => {
                const casterElem = elementById('casters');
                casterElem.innerHTML = '';
                nodecg.replicants.casters.value = {};

                casters.updateOrCreateCreateCasterElem(
                    'newcaster', { name: 'new caster', twitter: '@casternew', pronouns: 'she/her' }, true);

                const removeButton = elementById('remove-caster_newcaster');
                dispatch(removeButton, 'click');

                expect(casterElem.innerHTML).toEqual('');
            });
        });

        it('updates element with data if it exists', () => {
            const casterElem = elementById('casters');
            casterElem.innerHTML = `
                <div id="caster-container_newcaster">
                    <input id="caster-name-input_newcaster" class="test-input">
                    <input id="caster-twitter-input_newcaster" class="test-input">
                    <input id="caster-pronoun-input_newcaster" class="test-input">
                </div>`;

            casters.updateOrCreateCreateCasterElem(
                'newcaster', { name: 'new caster', twitter: '@casternew', pronouns: 'she/her' }, true);

            const casterNameInput = elementById<HTMLInputElement>('caster-name-input_newcaster');
            expect(casterNameInput.value).toEqual('new caster');
            expect(casterNameInput.classList).toContain('test-input');
            const casterTwitterInput = elementById<HTMLInputElement>('caster-twitter-input_newcaster');
            expect(casterTwitterInput.value).toEqual('@casternew');
            expect(casterTwitterInput.classList).toContain('test-input');
            const casterPronounInput = elementById<HTMLInputElement>('caster-pronoun-input_newcaster');
            expect(casterPronounInput.value).toEqual('she/her');
            expect(casterPronounInput.classList).toContain('test-input');
        });
    });

    describe('setUncommittedButtonDisabled', () => {
        it('disables appropriate buttons if disabled is true', () => {
            const buttonToDisable = document.createElement('button');
            buttonToDisable.classList.add('update-button', 'uncommitted');
            const buttonToKeepEnabled1 = document.createElement('button');
            buttonToKeepEnabled1.classList.add('update-button');
            const buttonToKeepEnabled2 = document.createElement('button');
            appendChildren(document.body, buttonToDisable, buttonToKeepEnabled1, buttonToKeepEnabled2);

            casters.setUncommittedButtonDisabled(true);

            expect(buttonToDisable.disabled).toEqual(true);
            expect(buttonToKeepEnabled1.disabled).toEqual(false);
            expect(buttonToKeepEnabled2.disabled).toEqual(false);
        });

        it('enables appropriate buttons if disabled is false', () => {
            const buttonToEnable = document.createElement('button');
            buttonToEnable.disabled = true;
            buttonToEnable.classList.add('update-button', 'uncommitted');
            const buttonToKeepDisabled1 = document.createElement('button');
            buttonToKeepDisabled1.classList.add('update-button');
            buttonToKeepDisabled1.disabled = true;
            const buttonToKeepDisabled2 = document.createElement('button');
            buttonToKeepDisabled2.disabled = true;
            appendChildren(document.body, buttonToEnable, buttonToKeepDisabled1, buttonToKeepDisabled2);

            casters.setUncommittedButtonDisabled(false);

            expect(buttonToEnable.disabled).toEqual(false);
            expect(buttonToKeepDisabled1.disabled).toEqual(true);
            expect(buttonToKeepDisabled2.disabled).toEqual(true);
        });
    });
});
