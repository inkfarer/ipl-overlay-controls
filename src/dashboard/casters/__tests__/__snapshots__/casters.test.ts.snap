// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`casters casters: change adds casters 1`] = `
"<div class=\\"space caster-container\\" id=\\"caster-container_castertwo\\">
    <div class=\\"select-container\\">
        <label for=\\"caster-name-input_castertwo\\">Name</label>
        <input type=\\"text\\" id=\\"caster-name-input_castertwo\\">
    </div>
    <div class=\\"layout horizontal select-container\\">
        <div class=\\"select-container caster-twitter-input-container\\">
            <label for=\\"caster-twitter-input_castertwo\\">Twitter</label>
            <input type=\\"text\\" id=\\"caster-twitter-input_castertwo\\">
        </div>
        <div class=\\"select-container\\">
            <label for=\\"caster-pronoun-input_castertwo\\">Pronouns</label>
            <input type=\\"text\\" id=\\"caster-pronoun-input_castertwo\\">
        </div>
    </div>
    <div class=\\"layout horizontal\\">
        <button class=\\"update-button max-width\\" id=\\"update-caster_castertwo\\" style=\\"background-color: var(--blue)\\">
            update
        </button>
        <button class=\\"red max-width\\" id=\\"remove-caster_castertwo\\">remove</button>
    </div></div><div class=\\"space caster-container\\" id=\\"caster-container_casterone\\">
    <div class=\\"select-container\\">
        <label for=\\"caster-name-input_casterone\\">Name</label>
        <input type=\\"text\\" id=\\"caster-name-input_casterone\\">
    </div>
    <div class=\\"layout horizontal select-container\\">
        <div class=\\"select-container caster-twitter-input-container\\">
            <label for=\\"caster-twitter-input_casterone\\">Twitter</label>
            <input type=\\"text\\" id=\\"caster-twitter-input_casterone\\">
        </div>
        <div class=\\"select-container\\">
            <label for=\\"caster-pronoun-input_casterone\\">Pronouns</label>
            <input type=\\"text\\" id=\\"caster-pronoun-input_casterone\\">
        </div>
    </div>
    <div class=\\"layout horizontal\\">
        <button class=\\"update-button max-width\\" id=\\"update-caster_casterone\\" style=\\"background-color: var(--blue)\\">
            update
        </button>
        <button class=\\"red max-width\\" id=\\"remove-caster_casterone\\">remove</button>
    </div></div>"
`;

exports[`casters casters: change handles casters being added and updated 1`] = `
"<div class=\\"space caster-container\\" id=\\"caster-container_castertwo\\">
    <div class=\\"select-container\\">
        <label for=\\"caster-name-input_castertwo\\">Name</label>
        <input type=\\"text\\" id=\\"caster-name-input_castertwo\\">
    </div>
    <div class=\\"layout horizontal select-container\\">
        <div class=\\"select-container caster-twitter-input-container\\">
            <label for=\\"caster-twitter-input_castertwo\\">Twitter</label>
            <input type=\\"text\\" id=\\"caster-twitter-input_castertwo\\">
        </div>
        <div class=\\"select-container\\">
            <label for=\\"caster-pronoun-input_castertwo\\">Pronouns</label>
            <input type=\\"text\\" id=\\"caster-pronoun-input_castertwo\\">
        </div>
    </div>
    <div class=\\"layout horizontal\\">
        <button class=\\"update-button max-width\\" id=\\"update-caster_castertwo\\" style=\\"background-color: var(--blue)\\">
            update
        </button>
        <button class=\\"red max-width\\" id=\\"remove-caster_castertwo\\">remove</button>
    </div></div><div class=\\"space caster-container\\" id=\\"caster-container_casterone\\">
    <div class=\\"select-container\\">
        <label for=\\"caster-name-input_casterone\\">Name</label>
        <input type=\\"text\\" id=\\"caster-name-input_casterone\\">
    </div>
    <div class=\\"layout horizontal select-container\\">
        <div class=\\"select-container caster-twitter-input-container\\">
            <label for=\\"caster-twitter-input_casterone\\">Twitter</label>
            <input type=\\"text\\" id=\\"caster-twitter-input_casterone\\">
        </div>
        <div class=\\"select-container\\">
            <label for=\\"caster-pronoun-input_casterone\\">Pronouns</label>
            <input type=\\"text\\" id=\\"caster-pronoun-input_casterone\\">
        </div>
    </div>
    <div class=\\"layout horizontal\\">
        <button class=\\"update-button max-width\\" id=\\"update-caster_casterone\\" style=\\"background-color: var(--blue)\\">
            update
        </button>
        <button class=\\"red max-width\\" id=\\"remove-caster_casterone\\">remove</button>
    </div></div>"
`;

exports[`casters casters: change handles casters being added and updated 2`] = `
"<div class=\\"space caster-container\\" id=\\"caster-container_casterthree\\">
    <div class=\\"select-container\\">
        <label for=\\"caster-name-input_casterthree\\">Name</label>
        <input type=\\"text\\" id=\\"caster-name-input_casterthree\\">
    </div>
    <div class=\\"layout horizontal select-container\\">
        <div class=\\"select-container caster-twitter-input-container\\">
            <label for=\\"caster-twitter-input_casterthree\\">Twitter</label>
            <input type=\\"text\\" id=\\"caster-twitter-input_casterthree\\">
        </div>
        <div class=\\"select-container\\">
            <label for=\\"caster-pronoun-input_casterthree\\">Pronouns</label>
            <input type=\\"text\\" id=\\"caster-pronoun-input_casterthree\\">
        </div>
    </div>
    <div class=\\"layout horizontal\\">
        <button class=\\"update-button max-width\\" id=\\"update-caster_casterthree\\" style=\\"background-color: var(--blue)\\">
            update
        </button>
        <button class=\\"red max-width\\" id=\\"remove-caster_casterthree\\">remove</button>
    </div></div><div class=\\"space caster-container\\" id=\\"caster-container_castertwo\\">
    <div class=\\"select-container\\">
        <label for=\\"caster-name-input_castertwo\\">Name</label>
        <input type=\\"text\\" id=\\"caster-name-input_castertwo\\">
    </div>
    <div class=\\"layout horizontal select-container\\">
        <div class=\\"select-container caster-twitter-input-container\\">
            <label for=\\"caster-twitter-input_castertwo\\">Twitter</label>
            <input type=\\"text\\" id=\\"caster-twitter-input_castertwo\\">
        </div>
        <div class=\\"select-container\\">
            <label for=\\"caster-pronoun-input_castertwo\\">Pronouns</label>
            <input type=\\"text\\" id=\\"caster-pronoun-input_castertwo\\">
        </div>
    </div>
    <div class=\\"layout horizontal\\">
        <button class=\\"update-button max-width\\" id=\\"update-caster_castertwo\\" style=\\"background-color: var(--blue)\\">
            update
        </button>
        <button class=\\"red max-width\\" id=\\"remove-caster_castertwo\\">remove</button>
    </div></div><div class=\\"space caster-container\\" id=\\"caster-container_casterone\\">
    <div class=\\"select-container\\">
        <label for=\\"caster-name-input_casterone\\">Name</label>
        <input type=\\"text\\" id=\\"caster-name-input_casterone\\">
    </div>
    <div class=\\"layout horizontal select-container\\">
        <div class=\\"select-container caster-twitter-input-container\\">
            <label for=\\"caster-twitter-input_casterone\\">Twitter</label>
            <input type=\\"text\\" id=\\"caster-twitter-input_casterone\\">
        </div>
        <div class=\\"select-container\\">
            <label for=\\"caster-pronoun-input_casterone\\">Pronouns</label>
            <input type=\\"text\\" id=\\"caster-pronoun-input_casterone\\">
        </div>
    </div>
    <div class=\\"layout horizontal\\">
        <button class=\\"update-button max-width\\" id=\\"update-caster_casterone\\" style=\\"background-color: var(--blue)\\">
            update
        </button>
        <button class=\\"red max-width\\" id=\\"remove-caster_casterone\\">remove</button>
    </div></div>"
`;

exports[`casters casters: change handles casters being removed 1`] = `
"<div class=\\"space caster-container\\" id=\\"caster-container_castertwo\\">
    <div class=\\"select-container\\">
        <label for=\\"caster-name-input_castertwo\\">Name</label>
        <input type=\\"text\\" id=\\"caster-name-input_castertwo\\">
    </div>
    <div class=\\"layout horizontal select-container\\">
        <div class=\\"select-container caster-twitter-input-container\\">
            <label for=\\"caster-twitter-input_castertwo\\">Twitter</label>
            <input type=\\"text\\" id=\\"caster-twitter-input_castertwo\\">
        </div>
        <div class=\\"select-container\\">
            <label for=\\"caster-pronoun-input_castertwo\\">Pronouns</label>
            <input type=\\"text\\" id=\\"caster-pronoun-input_castertwo\\">
        </div>
    </div>
    <div class=\\"layout horizontal\\">
        <button class=\\"update-button max-width\\" id=\\"update-caster_castertwo\\" style=\\"background-color: var(--blue)\\">
            update
        </button>
        <button class=\\"red max-width\\" id=\\"remove-caster_castertwo\\">remove</button>
    </div></div><div class=\\"space caster-container\\" id=\\"caster-container_casterone\\">
    <div class=\\"select-container\\">
        <label for=\\"caster-name-input_casterone\\">Name</label>
        <input type=\\"text\\" id=\\"caster-name-input_casterone\\">
    </div>
    <div class=\\"layout horizontal select-container\\">
        <div class=\\"select-container caster-twitter-input-container\\">
            <label for=\\"caster-twitter-input_casterone\\">Twitter</label>
            <input type=\\"text\\" id=\\"caster-twitter-input_casterone\\">
        </div>
        <div class=\\"select-container\\">
            <label for=\\"caster-pronoun-input_casterone\\">Pronouns</label>
            <input type=\\"text\\" id=\\"caster-pronoun-input_casterone\\">
        </div>
    </div>
    <div class=\\"layout horizontal\\">
        <button class=\\"update-button max-width\\" id=\\"update-caster_casterone\\" style=\\"background-color: var(--blue)\\">
            update
        </button>
        <button class=\\"red max-width\\" id=\\"remove-caster_casterone\\">remove</button>
    </div></div>"
`;

exports[`casters casters: change handles casters being removed 2`] = `
"<div class=\\"space caster-container\\" id=\\"caster-container_casterone\\">
    <div class=\\"select-container\\">
        <label for=\\"caster-name-input_casterone\\">Name</label>
        <input type=\\"text\\" id=\\"caster-name-input_casterone\\">
    </div>
    <div class=\\"layout horizontal select-container\\">
        <div class=\\"select-container caster-twitter-input-container\\">
            <label for=\\"caster-twitter-input_casterone\\">Twitter</label>
            <input type=\\"text\\" id=\\"caster-twitter-input_casterone\\">
        </div>
        <div class=\\"select-container\\">
            <label for=\\"caster-pronoun-input_casterone\\">Pronouns</label>
            <input type=\\"text\\" id=\\"caster-pronoun-input_casterone\\">
        </div>
    </div>
    <div class=\\"layout horizontal\\">
        <button class=\\"update-button max-width\\" id=\\"update-caster_casterone\\" style=\\"background-color: var(--blue)\\">
            update
        </button>
        <button class=\\"red max-width\\" id=\\"remove-caster_casterone\\">remove</button>
    </div></div>"
`;
